File: 2024/day20/go/solution.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

type Point struct {
	r, c int
}

func parseGrid(lines []string) ([][]byte, Point, Point) {
	grid := make([][]byte, len(lines))
	var start, end Point

	for r, line := range lines {
		grid[r] = []byte(line)
		for c, ch := range grid[r] {
			if ch == 'S' {
				start = Point{r, c}
			} else if ch == 'E' {
				end = Point{r, c}
			}
		}
	}

	return grid, start, end
}

func tracePath(grid [][]byte, start, end Point) map[Point]int {
	rows := len(grid)
	cols := len(grid[0])
	dist := make(map[Point]int)
	dist[start] = 0

	queue := []Point{start}
	directions := []Point{{-1, 0}, {1, 0}, {0, -1}, {0, 1}}

	for len(queue) > 0 {
		cur := queue[0]
		queue = queue[1:]

		if cur == end {
			break
		}

		for _, d := range directions {
			nr, nc := cur.r+d.r, cur.c+d.c
			next := Point{nr, nc}

			if nr >= 0 && nr < rows && nc >= 0 && nc < cols && grid[nr][nc] != '#' {
				if _, exists := dist[next]; !exists {
					dist[next] = dist[cur] + 1
					queue = append(queue, next)
				}
			}
		}
	}

	return dist
}

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

func countCheats(dist map[Point]int, maxCheatTime, minSavings int) int {
	count := 0

	// Convert map to slice for iteration
	trackPositions := make([]Point, 0, len(dist))
	for p := range dist {
		trackPositions = append(trackPositions, p)
	}

	for _, p1 := range trackPositions {
		d1 := dist[p1]
		for _, p2 := range trackPositions {
			// Manhattan distance is the cheat cost
			cheatCost := abs(p2.r-p1.r) + abs(p2.c-p1.c)
			if cheatCost <= maxCheatTime {
				d2 := dist[p2]
				savings := d2 - d1 - cheatCost
				if savings >= minSavings {
					count++
				}
			}
		}
	}

	return count
}

func part1(grid [][]byte, start, end Point) int {
	dist := tracePath(grid, start, end)
	return countCheats(dist, 2, 100)
}

func part2(grid [][]byte, start, end Point) int {
	dist := tracePath(grid, start, end)
	return countCheats(dist, 20, 100)
}

func main() {
	file, err := os.Open("../input.txt")
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error opening file: %v\n", err)
		os.Exit(1)
	}
	defer file.Close()

	var lines []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		if len(line) > 0 {
			lines = append(lines, line)
		}
	}

	if err := scanner.Err(); err != nil {
		fmt.Fprintf(os.Stderr, "Error reading file: %v\n", err)
		os.Exit(1)
	}

	grid, start, end := parseGrid(lines)

	fmt.Println("Part 1:", part1(grid, start, end))
	fmt.Println("Part 2:", part2(grid, start, end))
}

File: 2024/day20/php/solution.php
<?php

function parseGrid(string $input): array {
    $grid = [];
    $start = $end = null;
    $lines = explode("\n", trim($input));

    foreach ($lines as $r => $line) {
        $row = str_split($line);
        $grid[] = $row;
        foreach ($row as $c => $ch) {
            if ($ch === 'S') {
                $start = [$r, $c];
            } elseif ($ch === 'E') {
                $end = [$r, $c];
            }
        }
    }

    return [$grid, $start, $end];
}

function tracePath(array $grid, array $start, array $end): array {
    $rows = count($grid);
    $cols = count($grid[0]);
    $dist = [];
    $key = $start[0] . ',' . $start[1];
    $dist[$key] = 0;
    $queue = [$start];
    $front = 0;

    $directions = [[-1, 0], [1, 0], [0, -1], [0, 1]];

    while ($front < count($queue)) {
        [$r, $c] = $queue[$front++];

        if ($r === $end[0] && $c === $end[1]) {
            break;
        }

        $currentKey = $r . ',' . $c;

        foreach ($directions as [$dr, $dc]) {
            $nr = $r + $dr;
            $nc = $c + $dc;

            if ($nr >= 0 && $nr < $rows && $nc >= 0 && $nc < $cols && $grid[$nr][$nc] !== '#') {
                $nkey = $nr . ',' . $nc;
                if (!isset($dist[$nkey])) {
                    $dist[$nkey] = $dist[$currentKey] + 1;
                    $queue[] = [$nr, $nc];
                }
            }
        }
    }

    return $dist;
}

function countCheats(array $dist, int $maxCheatTime, int $minSavings): int {
    $count = 0;
    $trackPositions = [];

    // Convert dist keys back to coordinates
    foreach ($dist as $key => $d) {
        $parts = explode(',', $key);
        $trackPositions[] = [(int)$parts[0], (int)$parts[1]];
    }

    foreach ($trackPositions as [$r1, $c1]) {
        $d1 = $dist[$r1 . ',' . $c1];

        foreach ($trackPositions as [$r2, $c2]) {
            // Manhattan distance is the cheat cost
            $cheatCost = abs($r2 - $r1) + abs($c2 - $c1);

            if ($cheatCost <= $maxCheatTime) {
                $d2 = $dist[$r2 . ',' . $c2];
                $savings = $d2 - $d1 - $cheatCost;

                if ($savings >= $minSavings) {
                    $count++;
                }
            }
        }
    }

    return $count;
}

function part1(array $grid, array $start, array $end): int {
    $dist = tracePath($grid, $start, $end);
    return countCheats($dist, 2, 100);
}

function part2(array $grid, array $start, array $end): int {
    $dist = tracePath($grid, $start, $end);
    return countCheats($dist, 20, 100);
}

function main(): void {
    $input = file_get_contents(__DIR__ . '/../input.txt');
    [$grid, $start, $end] = parseGrid($input);

    echo 'Part 1: ' . part1($grid, $start, $end) . PHP_EOL;
    echo 'Part 2: ' . part2($grid, $start, $end) . PHP_EOL;
}

main();
